### PR TITLE
SR-7732: Dynamic casting CFError to Error results in a memory leak

### PR DESCRIPTION
**Pull request by tbkka**
The special handling for casting CFError/NSError to Swift Error
type was using cleanup code that didn't correctly handle this case.
This replaces the cleanup code with a more targeted version.

Resolves SR-7732
Resolves rdar://problem/40423061

### CODE REVIEW

**Review by DougGregor — APPROVED** [review]
Thank you! Please cherry-pick to the swift-5.2-branch

File: test/stdlib/ErrorBridged.swift [hunk 1]
```diff
@@ -788,4 +788,23 @@ ErrorBridgingTests.test("error-to-NSObject casts") {
...
 }
 
+// SR-7732: Casting CFError to Error results in a memory leak
+ErrorBridgingTests.test("CFError-to-Error casts") {
+  func should_not_leak() {
+    let something: Any? = NSError(domain: "Foo", code: 1)
+
+    if let error = something as? Error {
+      expectEqual(
+        "The operation couldn’t be completed. (Foo error 1.)",
```
Comment (benrimmington) [inline test/stdlib/ErrorBridged.swift]: `"The operation couldn\\U2019t be completed. (%@ error %ld.)"` is [localized][] in CoreFoundation. So will this test fail when the user's preferred language isn't English?

[localized]: <https://github.com/apple/swift-corelibs-foundation/blob/653c070d8ee571611ab4cc3e63d640ef4b8d151c/CoreFoundation/Error.subproj/CFError.c#L307>

Comment (tbkka) [reply test/stdlib/ErrorBridged.swift]: Good catch!  I did actually realize that would be a problem when I wrote it, but forgot to go back and find a better approach.  Thanks for pointing it out!

**Comment by tbkka** [comment]
@swift-ci Please test macOS

**Comment by tbkka** [comment]
@swift-ci Please smoke test Linux

**Comment by tbkka** [comment]
@swift-ci Please smoke test

**Comment by tbkka** [comment]
@swift-ci Please smoke test macOS

File: test/stdlib/ErrorBridged.swift [hunk 2]
```diff
@@ -788,4 +788,23 @@ ErrorBridgingTests.test("error-to-NSObject casts") {
   }
 }
 
+// SR-7732: Casting CFError to Error results in a memory leak
+ErrorBridgingTests.test("CFError-to-Error casts") {
+  func should_not_leak() {
+    let something: Any? = NSError(domain: "Foo", code: 1)
```
[the remark this answers is not in the document]
Comment (tbkka) [reply test/stdlib/ErrorBridged.swift]: I added a separate test to explicitly verify CFError.

**Comment by tbkka** [comment]
@swift-ci Please smoke test Linux

**Comment by tbkka** [comment]
@swift-ci Please smoke test macOS

File: test/stdlib/ErrorBridged.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikeash) [reply test/stdlib/ErrorBridged.swift]: Yay!

**Review by mikeash — APPROVED** [review]

**Pull request merged**